src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java

```diff
@@ -0,0 +1,200 @@
+            .getInputParameters()
+            .get(CREATE_NOTEBOOK_PARAMETERS, ApiGcpAiNotebookInstanceCreationParameters.class);
+    setFields(creationParameters, instance);
+
+    String serviceAccountEmail =
+        CreateServiceAccountStep.serviceAccountEmail(
+            flightContext.getWorkingMap().get(CREATE_NOTEBOOK_SERVICE_ACCOUNT_ID, String.class),
+            projectId);
+    instance.setServiceAccount(serviceAccountEmail);
+
+    // Create the AI Notebook instance in the service account proxy mode to control proxy access by
+    // means of IAM permissions on the service account.
+    // https://cloud.google.com/ai-platform/notebooks/docs/troubleshooting#opening_a_notebook_results_in_a_403_forbidden_error
+    ImmutableMap<String, String> metadata =
+        new ImmutableMap.Builder<String, String>().put("proxy-mode", "service_account").build();
+    instance.setMetadata(metadata);
+
+    setNetworks(instance, projectId, creationParameters.getLocation());
+
+    return instance;
+  }
+
+  private static void setFields(
+      ApiGcpAiNotebookInstanceCreationParameters creationParameters, Instance instance) {
+    instance
+        .setPostStartupScript(creationParameters.getPostStartupScript())
+        .setMachineType(creationParameters.getMachineType());
+    ApiGcpAiNotebookInstanceVmImage vmImageParameters = creationParameters.getVmImage();
+    if (vmImageParameters != null) {
+      instance.setVmImage(
+          new VmImage()
+              .setProject(vmImageParameters.getProjectId())
+              .setImageFamily(vmImageParameters.getImageFamily())
+              .setImageName(vmImageParameters.getImageName()));
+    }
+    ApiGcpAiNotebookInstanceContainerImage containerImageParameters =
+        creationParameters.getContainerImage();
+    if (containerImageParameters != null) {
+      instance.setContainerImage(
+          new ContainerImage()
+              .setRepository(containerImageParameters.getRepository())
+              .setTag(containerImageParameters.getTag()));
+    }
+  }
+
+  private static void setNetworks(Instance instance, String projectId, String location) {
+    // 'network' is the name of the VPC network instance created by the Buffer Service.
+    // TODO(PPF-469): Instead of hard coding this, look up the name of the network on the project.
+    instance.setNetwork("projects/" + projectId + "/global/networks/network");
+    // Assume the zone is related to the location like 'us-west1' is to 'us-west1-b'.
+    Preconditions.checkArgument(location.length() > 2, "Invalid location '%s'", location);
+    String zone = location.substring(0, location.length() - 2);
+    // Like 'network', 'subnetwork' is the name of the subnetwork created by the Buffer Service in
+    // each zone.
+    instance.setSubnet("projects/" + projectId + "/regions/" + zone + "/subnetworks/subnetwork");
+  }
+
+  private InstanceName createInstanceName(String projectId) {
+    return InstanceName.builder()
```

Comment:
What if we have already created an instance of this name in this workspace?
Do we need another step that looks in the workspace metadata to make sure that is not the case?

Reply:
per other comment, I was relying on Jay's PR #291 to do the resource uniqueness check at ResourceDao insertion time (the earlier `StoreMetadataStep` in this flight). I agree that without that this is a problem. I was not planning on blocking this PR on that, but if that doesn't get merged we'll have to come up with another solution here.
This is a generic problem for controlled resources and their underlying cloud resource.

Reply:
OK, that sounds fine to me